sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2670,6 +2670,7 @@ class Dataset[T] private[sql](
    *   <li>min</li>
    *   <li>max</li>
    *   <li>arbitrary approximate percentiles specified as a percentage (e.g. 75%)</li>
+   *   <li>count_distinct</li>
```

Comment:
I am not sure how common it is to show distinct count. I believe this method was inspired by [`pandas.DataFrame.describe()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) which however does not include distinct count.

BTW, I think you should update the examples below too.

Reply:
I wouldn't add this if there's any strong reason for it because you're already able to do this via `agg`.

Reply:
Meh, I think this could be useful as an _option_; it doesn't really expand the API surface or complexity. I'm OK with adding it, FWIW.

Reply:
Wouldn't be more useful to expose approx variants here?

Reply:
@zero323 - I updated the PR to expose `approx_count_distinct` as well, good point.

[One SO question on this](https://stackoverflow.com/questions/37949494/how-to-count-occurrences-of-each-distinct-value-for-every-column-in-a-dataframe) has 81,000 views and [the other](https://stackoverflow.com/questions/40888946/spark-dataframe-count-distinct-values-of-every-column) has 66,000 views.  There is a high demand for count distinct / approx_count_distinct exploratory data analysis queries by columns among the Spark userbase.

These options don't cost us much and should help the less technical Spark users that aren't comfortable using `agg`.  I help lots of folks that can't even program and they aren't code geniuses like you all ;)

Reply:
I am not sure. You can all have what you want via `agg`. This is just a shortcut to show basic stats, which was inspired by pandas' one, see https://github.com/apache/spark/commit/e1a172c201d68406faa53b113518b10c879f1ff6. We should of course avoid showing every stats here.

You can easily do `df.agg("col" -> "approx_count_distinct", ...)` easily for other stats. Adding `count_distinct` here makes it ambiguous what to add in `DataFrame.describe` and `DataFrame.summary`.

Reply:
It's true, though, users looking for summary stats are probably looking at summary(), and sometimes looking for distinct values, and this puts them into one output with the others. I think it's slightly worth adding one more somewhat-common summary stat.

Reply:
> I am not sure. You can all have what you want via `agg`

Which is also more suitable for anything but quick visual inspection.

My biggest concern is somewhat different though ‒ existing summaries have good performance and predictable behavior. In contrast count distinct can be unstable and approx count distinct slower and more resource hungry.

Reply:
Sure, but this is only computed if the caller requests it.

Reply:
@HyukjinKwon - you raise a good point about API ambiguity between `Dataset.describe` and `Dataset.summary`, but those ambiguities already exist.  `df.describe("some_col")`, `df.select("some_col").summary()`, and `df.select("some_col").summary("37%", "87%")` all return different results.

`describe` is a "partial alias" of `summary`, but the method signatures are different.

The Spark & Pandas `describe` methods aren't consistent either.  The Spark `describe` method takes column name arguments whereas the Pandas `describe` method takes options, see the Pandas method signature: `DataFrame.describe(percentiles=None, include=None, exclude=None, datetime_is_numeric=False)`.

A Pandas user is already going to be confused by the API.  They're going to run `df.describe(percentiles=Array(.25, .5, .75))` and have to Google around to find that they need to run `df.summary("0.25%", "0.5%", "0.75%")`.  

Adding additional options to `Dataset.summary` seems consistent with the existing design.  The `summary` method is already where we put the extra metrics that aren't enabled by default, like the option to add arbitrary percentiles.

BTW - I don't mean to complain here ;). It's easy for me to make comments on API decisions made years ago from folks in the trenches.  You are all doing a wonderful job maintaining this project and continuously moving it in a good direction.  I'm always impressed by the code you write and will hopefully get to your level some day.  My PR comments sound critical cause I'm just trying to get to the point.  I'm very grateful for all the work you've done on this project.

Reply:
I would favor this change. If there aren't significant objections, I'd like to merge it.

Reply:
I'm okay given that we don't change the default.

Reply:
Sounds good.